File: src/contest/414/Q2.ts
export function maxPossibleScore(start: number[], d: number): number {
    start.sort((a, b) => a - b);
    const n = start.length;
    const check = (mid: number) => {
        let last = start[0];
        for (let i = 1; i < n; i++) {
            const rS = start[i];
            const rE = start[i] + d;

            if (last + mid <= rE) {
                last = Math.max(last + mid, rS);
            } else {
                return false;
            }
        }
        return true;
    };

    let res = 0;
    let l = 0;
    let r = start[n - 1] + d - start[0];

    while (l <= r) {
        const mid = l + Math.floor((r - l) / 2);
        if (check(mid)) {
            res = mid;
            l = mid + 1;
        } else {
            r = mid - 1;
        }
    }

    return res;
}

File: src/contest/414/Q2.go
package c_414

import "sort"

func MaxPossibleScore(start []int, d int) int {
	n := len(start)

	sort.Slice(start, func(i, j int) bool {
		return start[i] < start[j]
	})

	l := 0
	r := start[n-1] + d - start[0]
	res := 0

	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}

	canAchieve := func(mid int) bool {
		last := start[0]
		for i := 1; i < n; i++ {
			rS := start[i]
			rE := start[i] + d

			if last+mid <= rE {
				last = max(last+mid, rS)
			} else {
				return false
			}
		}
		return true
	}

	for l <= r {
		mid := l + (r-l)/2

		if canAchieve(mid) {
			res = mid
			l = mid + 1
		} else {
			r = mid - 1
		}
	}

	return res
}
